timemachine/potentials/bonded.py

```diff
@@ -147,6 +151,95 @@ def harmonic_bond(conf, params, box, lamb, bond_idxs, lamb_mult=None, lamb_offse
     return jnp.sum(energy)
 
 
+def get_centroid_cos_angles(conf, angle_idxs):
+
+    vij = []
+    vik = []
+
+    for idxs, j, k in angle_idxs:
+
+        x_j = conf[j]
+        # compute v-site from unit vectors relative to i
+        if len(idxs) == 2:
+            a, b = idxs
+            x_a, x_b = conf[a], conf[b]
+            x_ia = x_a - x_j
+            x_ia = x_j + x_ia / jnp.linalg.norm(x_ia)
+            x_ib = x_b - x_j
+            x_ib = x_j + x_ib / jnp.linalg.norm(x_ib)
+            x_i = (x_ia + x_ib) / 2.0
+        elif len(idxs) == 3:
+            a, b, c = idxs
+            x_a, x_b, x_c = conf[a], conf[b], conf[c]
+            x_ia = x_a - x_j
+            x_ia = x_j + x_ia / jnp.linalg.norm(x_ia)
+            x_ib = x_b - x_j
+            x_ib = x_j + x_ib / jnp.linalg.norm(x_ib)
+            x_ic = x_c - x_j
+            x_ic = x_j + x_ic / jnp.linalg.norm(x_ic)
+            x_i = (x_ia + x_ib + x_ic) / 3.0
+        else:
+            assert 0
+
+        vij.append(x_i - x_j)
+        x_k = conf[k]
+        vik.append(x_k - x_j)
+
+    vij = jnp.array(vij)
+    vik = jnp.array(vik)
+
+    top = jnp.sum(jnp.multiply(vij, vik), -1)
+    bot = jnp.linalg.norm(vij, axis=-1) * jnp.linalg.norm(vik, axis=-1)
+
+    tb = top / bot
```

Comment:
Extract function to compute `cos_angle(x_a, x_b, x_c)`, reuse here and in `harmonic_angle`?

(So that responsibility of `get_centroid_cos_angles` can be simplified to just defining the centroid `x_i`)

Reply:
Will refactor!

Reply:
refactored in https://github.com/proteneer/timemachine/pull/746/commits/07eeaa154e1bba54fc1c7a8512410470aef8709f